### PR TITLE
.NET Core 3.1 + NET 5 Support

### PR DESCRIPTION
**Pull request by 304NotModified**
resolves https://github.com/NLog/NLog.Windows.Forms/issues/39

works locally in test application

todo:

- [x] Add .NET Core 3
- [x] Add .NET 5
- [x] fix clicking link in .NET 5
- [x] disabled SupportLinks for .NET Core 3, but not for .NET 5
- [x] run unit tests for all platforms
- [x] run unit tests for all platforms in CI (AppVeyor) -> started
- [x] fix issues in other platforms, see unit tests
- [x] disable the linkclick tests for .NET core 3
- [x] move to .NET Core 3.1 instead of 3.0

### CODE REVIEW

**Comment by 304NotModified** [comment]
hyperlinks won't work well

expected:

image


without change:

image


with change in PR (still no good)

image

**Comment by 304NotModified** [comment]
I think I did, as preview 8 is production ready according to ms

**Comment by Logerfo** [comment]
Take a look at dotnet/winforms#1631.

**Comment by 304NotModified** [comment]
ah thanks!

**Comment by Logerfo** [comment]
If that was indeed the problem, dotnet/winforms#1745 was merged today and will be available through the .NET Core 3.0 GA, which will be [released](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0-preview-9/) at .NET Conference later this month (between September 23th and 25th).

**Comment by 304NotModified** [comment]
Nice!

It isn't fixed in preview 9?

**Comment by Logerfo** [comment]
I don't think so, the merge happened a few hours later than the release announcement. Perhaps you could try it with the `master` "5.0" SDK, but I don't know how to do that.

**Comment by Logerfo** [comment]
I think that would help if you want to try: https://github.com/dotnet/coreclr#using-your-build

**Comment by Logerfo** [comment]
.NET Core 3.0 RC1 is up, did you have a chance to try it out yet?

**Comment by 304NotModified** [comment]
Will try to test soon. If you have time, please test it, the case is really easy to see (just start the netcore3 app). I'm now on holiday so a bit limited.

**Comment by Logerfo** [comment]
Running with RC1:
image

**Comment by Logerfo** [comment]
I think that an issue with a proper minimum reproduction should be opened at dotnet/winforms. I'm not going to do that because I'm not fully aware of stuff. I don't use this package with RTF. Actually, I don't use RTF at all, so this hyperlink syntax seems out of place for me. Nevertheless, this will not be fixed before the GA release.

**Comment by 304NotModified** [comment]
Unfortunately I don't know if it isn't a bug in our code or Microsoft's

**Comment by Logerfo** [comment]
Tested against .NET Core 3.1 Preview 1: same behavior.

**Comment by 304NotModified** [comment]
Thanks for testing. I think the issue is here:

https://github.com/NLog/NLog.Windows.Forms/blob/38ebb7d8c422958110c33024c5c51efc7defabd7/NLog.Windows.Forms/RichTextBoxTarget.cs#L835-L867

But dunno what the problem is there.

**Comment by Logerfo** [comment]
I did some digging, hope it helps.

## Predefined stuff
```js
linkAddRegex = /(\([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\))/g
textBox.Text = "Init link text in config\nstart thread (aab408af-e4bb-4f33-80d7-8ab6420b1d5c)\n"
```
## Actual code
```c#
textBox.SelectionStart = startIndex; //SelectionStart = 25
textBox.SelectionLength = textBox.Text.Length - textBox.SelectionStart; //SelectionLength = 77 - 24
string addedText = textBox.SelectedText; //addedText = "start thread (aab408af-e4bb-4f33-80d7-8ab6420b"
MatchCollection matches = linkAddRegex.Matches(addedText); //matches.Count = 0
for (int i = matches.Count - 1; i >= 0; --i) //never enters
```
## What is happening
- `77 - 24 = 53`.
- The selection `(25, 53)` gives `start thread (aab408af-e4bb-4f33-80d7-8ab6420b1d5c)`, which matches the regular expression.
- But, as you can see above, the actual selection text is `start thread (aab408af-e4bb-4f33-80d7-8ab6420b`, which would be a `(25, 47)` selection and doesn't match the regular expression.
- How? Because in .NET Core, `textBox.SelectionLength` is `47`, even after assigning `53` to it, while in .NET Framework, it works as expected.
- Why? No idea.

Heres the `SelectionLength` code: https://github.com/dotnet/winforms/blob/2e00da49a17d194634ef7291444cab4a5d9ddf38/src/System.Windows.Forms/src/System/Windows/Forms/RichTextBox.cs#L1242-L1274

I test it both against .NET Framework 4.8 (which works) and .NET Core 3.1 preview 1 (which doesn't). The property code is the same for both of them. The value of `IsHandleCreated` is `true` in both cases. The `SelectedText` property code also seems to be the same for both targets, but they're weirder, so I'm not sure.

**Comment by tomdewilde** [comment]
Is there any update on this issue ? I would like to convert some of my projects to .net core 3.1, but they are depending on your nlog.windows.forms package ... so I cannot switch as long as this is not solved. Thanks for an update !

**Comment by 304NotModified** [comment]
Do you need the RichTextBoxTarget? Maybe I should remove that for .net core 3, or maybe only the hyperlink feature

**Comment by tomdewilde** [comment]
Thank you for your reply Julian. Yes, I extensively use the
RichTextBoxTarget, but never used the hyperlink feature before and am not
planning to. I really appreciate your work on this !

**Comment by Logerfo** [comment]
FYI I'm not working with Windows Forms anymore, but I never used the `RichTextBoxTarget` at all when working with `NLog.Windows.Forms`. My use case was message boxes.

**Comment by kirsan31** [comment]
@304NotModified 

We defiantly need .net core (3.1) support, actually Net (5) already.

> hyperlinks won't work well

This is known [issue](https://github.com/dotnet/winforms/issues/3632).

> Do you need the RichTextBoxTarget? Maybe I should remove that for .net core 3, or maybe only the hyperlink feature

Yes we are actively using RichTextBoxTarget. I think the best workaround it's something disable hyperlink feature until it will be fixed in 6.0...

**Comment by 304NotModified** [comment]
> Yes, we are using RichTextBoxTarget. I think that the best workaround it's somehow disable hyperlink feature until it will be fixed in Net...

Agree with that. Any help on this would be great!

**Comment by weltkante** [comment]
Its not a regression, the RTF generated is using undocumented behavior, the RTF generator needs to be updated to generate something conforming to the RTF spec (its simple to test, just save the RTF to a file and open in wordpad, if it doesn't work its likely not according to spec). I have some suggestions how to fix it [over in the WinForms issue discussion](https://github.com/dotnet/winforms/issues/3632#issuecomment-752274890)

**Comment by 304NotModified** [comment]
@kirsan31 If I reboot this PR and drop the support for hyperlinks (for now), could you please test it? 

WinForms is a bit long ago for me

**Comment by kirsan31** [comment]
> @kirsan31 If I reboot this PR and drop the support for hyperlinks (for now), could you please test it?
> 
> WinForms is a bit long ago for me

Yes, of course.

**Comment by weltkante** [comment]
Ok I've downloaded the PR and played around with it, can confirm .NET 4.7.2 and .NET 3.x are broken - but .NET 5.0 actually works. Whatever the issue with the selection was has been fixed for .NET 5

However there is some bad news, the code uses the same broken hyperlink RTF as in the linked WinForms issue. Funny enough, contrary to the WinForms issue, it still renders in this PR. Something about your RTF must be different from the repro in the WinForms issue.

Either way, due to another bugfix your hack of piping hidden text through the LinkClicked handler no longer works, i.e. when clicking the link it won't see the hidden text in the event, thus the regex does not match and the click gets ignored.

TLDR:
- PR works in .NET 5
- PR does not use official Hyperlink RTF markup, but works anyways (but there are cases where it doesn't work, see dotnet/winforms#3632, its unclear what the differences are)
- Hyperlink clicks do not work because .NET Core fixed bugs in handling of hidden text, thereby changing the event args of the link click to no longer include hidden text.

I'm not sure if the last point is an intentional fix, I'll raise a WinForms issue about handling of hidden text in LinkClicked events for further discussion.

**Comment by 304NotModified** [comment]
There is some more work to be done here,

some features won't work, see the failing unit tests. We need also to disabled the linkclick tests for .NET core 3

image

**Comment by weltkante** [comment]
Seems some tests are overly strict at checking exact RTF snippets, there is some whitespace difference in the markup sections. Also [the default font differs](https://docs.microsoft.com/en-us/dotnet/core/compatibility/fx-core#default-control-font-changed-to-segoe-ui-9-pt) between Desktop Framework and .NET Core, so the tests are picking up on that difference. You either may want to adjust the tests, or explicitly set a font during form creation.

**Comment by weltkante** [comment]
MaxLines test fails because `ShowMinimized = true` and in this case the RichTextBox apparently has zero width when minimized. The MaxLines implementation is using the visual lines not logical lines, so a zero width textbox will perform weird line breaking which confuses the MaxLines implementation.

**Comment by 304NotModified** [comment]
Thanks a lot!

**Comment by weltkante** [comment]
I can get all tests to pass (except the link tests in cases which we expect to fail) with the following changes:
- break apart the `\viewkind4\uc1\pard\cf1...` asserts into two asserts because there can be whitespace after the second markup tag:
  ```csharp
  Assert.Contains(@"\viewkind4\uc1", result);
  Assert.Contains(@"\pard\cf1......\par", result);
  ```
- in `FormHelper.CreateForm` use `Font = new Font("Microsoft Sans Serif", 7.5f, FontStyle.Regular, GraphicsUnit.Point, 0),` to initialize the form (apparently thats what the tests were written against, if you use different parameters you have to update the tests)
- remove `ShowMinimized = true` from `MaxLinesTest` as explained above

**Comment by 304NotModified** [comment]
@weltkante thanks! Maybe it's easier to push to this branch :) I have invited you so you will have push access :)

**Comment by weltkante** [comment]
Actually I'm not sure how to do that, but I forked and pushed my remaining changes [here](https://github.com/NLog/NLog.Windows.Forms/compare/core3...weltkante:core3). Nothing speical really.

**Comment by weltkante** [comment]
Oh I see AppVeyor tests use FontSize 8.25 (fs17) instead of 7.5 (fs15) so having the font initialization set to a fixed font in CreateForm is disturbing those tests. I'm not really a fan of hardcoding fonts over the OS default anyways, but I'm not sure how to test that then. Any suggestion how to proceed? Remove the `if (IsAppVeyor)` conditionals from the test?

**Comment by weltkante** [comment]
[Here](https://github.com/NLog/NLog.Windows.Forms/compare/core3...weltkante:corefont) is what it would look like to not set a fixed font and instead branch the tests. You might have to add `if (IsAppVeyor)` conditionals into the .NET Core conditionals as well if font sizes differ here.

**Comment by weltkante** [comment]
Did a mistake in my original tests, the officially documented hyperlink markup actually works in .NET 5 (the native control does all the work, WinForms isn't involved) - switching the link generation to not use hidden text means we don't have to wait for a fix for WinForms to pass hidden text to the `LinkClicked` handler.

**Comment by kirsan31** [comment]
@304NotModified 

> move to .NET Core 3.1 instead of 3?

I think yes, only NET 5 and .NET Core 3.1 are actual.


@weltkante 
Thank you once more, your are the best, as always 😉

**Comment by 304NotModified** [comment]
> I'm not really a fan of hardcoding fonts over the OS default anyways, but I'm not sure how to test that then. Any suggestion how to proceed?

I think it would be nice to have fixed fonts for the tests, and only for the tests. 

>You mave have changed your default font size on your local machine?

No, but maybe the zoom factor is also related?

**Comment by weltkante** [comment]
Since I've already been halfway there I've also [ported](https://github.com/NLog/NLog.Windows.Forms/compare/core3...weltkante:hiddentext) the `RichTextBox.Text` accessor fix from .NET 5 (which returns hidden text instead of suppressing it and therefore messing up indexing) into a standalone method. This makes links generate properly in .NET Core 3.x as well I believe, but I didn't test it extensively, the tests certainly pass. Your choice if you want that additional interop for working around a .NET Core 3.x bug, or just declare it not supported.

Technically you can use that on .NET 4.7+ as well but considering that there is a runtime switch to decide which RTF control to use it would probably get pretty hacky to figure out which RTF control is used. Certainly not enough to have static `#if` compile time switching between the strategies on Desktop Framework, you have to actually check in which version the RTF control is running and then select dynamically.

**Comment by 304NotModified** [comment]
> I think yes, only NET 5 and .NET Core 3.1 are actual.

Ow yes indeed, EOL 2020-03-03. At work we are already at 3.1/5.0 😅

**Comment by weltkante** [comment]
> I think it would be nice to have fixed fonts for the tests, and only for the tests.

Any idea how to configure the font dynamically generated form? Would certainly simplify the tests a lot.

> but maybe the zoom factor is also related?

Perhaps, it certainly complicates the tests (I have to switch the tests to run with AppVeyor settings because whatever is the standard on your side isn't for me).

**Comment by 304NotModified** [comment]
> Any idea how to configure the font of the dynamically generated form? Would certainly simplify the tests a lot.

Well we could make it an option of the target (the font size) and then set a fixed from the tests?

**Comment by 304NotModified** [comment]
It took some time, but it's finally merged :) 

Thanks you guys!!

**Comment by 304NotModified** [comment]
It's published as 4.5.0-rc.1: https://www.nuget.org/packages/NLog.Windows.Forms/4.5.0-rc.1
